src/aspire/abinitio/commonline_sync3n.py

```diff
@@ -0,0 +1,1025 @@
+                logger.info("GPU not found, defaulting to numpy.")
+
+        except ModuleNotFoundError:
+            logger.info("cupy not found, defaulting numpy.")
+
+    ###########################################
+    # High level algorithm steps              #
+    ###########################################
+    def estimate_rotations(self):
+        """
+        Estimate rotation matrices.
+
+        :return: Array of rotation matrices, size n_imgx3x3.
+        """
+
+        # Initial estimate of viewing directions
+        Rijs0 = self._estimate_relative_viewing_directions()
+
+        # Compute and apply global handedness
+        Rijs = self._global_J_sync(Rijs0)
+
+        # Build sync3n matrix
+        S = self._construct_sync3n_matrix(Rijs)
+
+        # Optionally compute S weights
+        W = None
+        if self.S_weighting is True:
+            W = self._syncmatrix_weights(Rijs)
+
+        # Yield rotations from S
+        self.rotations = self._sync3n_S_to_rot(S, W)
+
+    #######################
+    # Main Sync3N Methods #
+    #######################
+    def _sync3n_S_to_rot(self, S, W=None, n_eigs=4):
+        """
+        Use eigen decomposition of S to estimate transforms,
+        then project transforms to nearest rotations.
+
+        :param S: Numpy array represeting Synchronization matrix.
+        :param W: Optional weights array, default `None` is equal weighting of `S`.
+        :param n_eigs: Optional, number of eigenvalues to compute (min 3).
+        """
+
+        if n_eigs < 3:
+            raise ValueError(
+                f"n_eigs must be greater than 3, default is 4. Invoked with {n_eigs}"
+            )
+
+        if W is not None:
+            logger.info("Applying weights to synchronization matrix.")
+            if not W.shape == (self.n_img, self.n_img):
+                raise RuntimeError(
+                    f"Shape of W should be {(self.n_img, self.n_img)}."
+                    f" Received {W.shape}."
+                )
+            # Initialize D
+            D = np.mean(W, axis=1)  # D, check axis
+
+            Dhalf = D
+            # Compute mask of trouble D values
+            nulls = np.abs(D) < self._D_null
+            # Avoid trouble values when exponentiating
+            Dhalf[~nulls] = Dhalf[~nulls] ** (-0.5)
+            # Flush trouble values to zero
+            Dhalf[nulls] = 0
+            # expand diagonal
+            Dhalf = np.diag(Dhalf)
+
+            # Report W Diagnostic
+            W_normalized = Dhalf**2 @ W
+            nzidx = np.sum(W_normalized, axis=1) != 0
+            err = np.linalg.norm(np.sum(W_normalized[nzidx], axis=1) - self.n_img)
+            if err > 1e-10:
+                logger.warning(f"Large Weights Matrix Normalization Error: {err}")
+
+            # Make W of size 3Nx3N
+            W = np.kron(W, np.ones((3, 3)))
+
+            # Make Dhalf of size 3Nx3N
+            Dhalf = np.diag(np.kron(np.diag(Dhalf), np.ones((1, 3)))[0])
```

Comment:
I'll take the first one, oversight on my part translating the code, but the second one is not the same...

```
>>> Dhalf = np.arange(9).reshape(3,3)
>>> a=np.diag(np.kron(np.diag(Dhalf), np.ones((1, 3)))[0])
>>> b=np.diag(np.kron(np.diag(Dhalf), np.ones(3)))
>>> c=np.kron(Dhalf, np.diag(np.ones(3)))
>>> 
>>> a
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
>>> b
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
>>> c
array([[0., 0., 0., 1., 0., 0., 2., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 2., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 2.],
       [3., 0., 0., 4., 0., 0., 5., 0., 0.],
       [0., 3., 0., 0., 4., 0., 0., 5., 0.],
       [0., 0., 3., 0., 0., 4., 0., 0., 5.],
       [6., 0., 0., 7., 0., 0., 8., 0., 0.],
       [0., 6., 0., 0., 7., 0., 0., 8., 0.],
       
 ```

Reply:
Sounds good. Third one does work for an already diagonal `Dhalf`.

Reply:
Oh, I see so `d=p.kron(np.diag(Dhalf), np.diag(np.ones(3))). I'll try it

Reply:
Nah, still didn't work. Going to stick with the one referred to as `b` above.  That removes the silly oversight I had in there. Thanks for catching.

Reply:
Sounds good!

But just to clarify, `Dhalf` in the code is already a diagonal matrix. Your example starts with a non-diagonal `Dhalf`. If you start with `Dhalf` diagonal `a`, `b`, and `c` should be the same.

Reply:
See above. I've tried both ways (c, then d), neither worked 🤷‍♂️ .

Reply:
```
>>> Dhalf = np.diag(np.diag(np.arange(9).reshape(3,3)))
>>> Dhalf
array([[0, 0, 0],
       [0, 4, 0],
       [0, 0, 8]])
>>> a=np.diag(np.kron(np.diag(Dhalf), np.ones((1, 3)))[0])
>>> b=np.diag(np.kron(np.diag(Dhalf), np.ones(3)))
>>> c=np.kron(Dhalf, np.diag(np.ones(3)))
>>> a
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
>>> b
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
>>> c
array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 4., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 4., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 4., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 8., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 8., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 8.]])
```

Reply:
I would be super caution about injecting implicit behavior here.
These are _clearly_ not doing the same thing outside a special case...  While it may be the one used in that line today, if any other line is changed, yikes.... let's not.

>>> Dhalf = np.arange(9).reshape(3,3)
>>> d=np.kron(np.diag(Dhalf), np.diag(np.ones(3)))                                             
>>> d                                                                                          
array([[0., 0., 0., 4., 0., 0., 8., 0., 0.],                                                   
       [0., 0., 0., 0., 4., 0., 0., 8., 0.],                                                   
       [0., 0., 0., 0., 0., 4., 0., 0., 8.]])

Reply:
fair enough!